### PR TITLE
[Integration Tests] Consensus reorg tests part 3

### PR DESCRIPTION
**Pull request by fassadlr**
With this test we introduce a block observer feature for integration tests which can catch when reorg happened and execute  a given func at a certain block height. Once the func has been executed it will be executed again. This can be changed at a later stage to make this more configurable etc.

### CODE REVIEW

File: src/Stratis.Bitcoin.IntegrationTests/ConsensusManagerInterceptorTests.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                {
+                    if (chainedHeaderBlock.ChainedHeader.Previous.Height == 5)
+                    {
+                        TestHelper.Disconnect(minerA, minerC);
+                        return true;
+                    }
+
+                    return false;
+                }
+                minerA.BlockDisconnectInterceptor(interceptor);
```
Comment (dangershony) [inline src/Stratis.Bitcoin.IntegrationTests/ConsensusManagerInterceptorTests.cs]: Would be cleaner if you did 
```
 var minerA = builder.CreateStratisPosNode(new StratisRegTest()).NotInIBD().WithWallet().BlockDisconnectInterceptor(interceptor).Start();
```

Or even just call it `Intercept(`

File: src/Stratis.Bitcoin.IntegrationTests/ConsensusManagerInterceptorTests.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                TestHelper.BuildBlocks.OnNode(minerB).Amount(5).Invalid(8, (node, block) => BlockBuilder.InvalidCoinbaseReward(node, block)).BuildAsync();
+
+                // Reconnect MinerA to MinerB.
+                TestHelper.Connect(minerA, minerB);
+
+                // MinerC should be disconnected from MinerA.
+                TestHelper.WaitLoop(() => !TestHelper.IsNodeConnectedTo(minerA, minerC));
+
+                // This will cause the reorg chain to fail at block 8 and roll back any changes.
+                TestHelper.WaitLoop(() => TestHelper.IsNodeSyncedAtHeight(minerA, 9));
```
Comment (dangershony) [inline src/Stratis.Bitcoin.IntegrationTests/ConsensusManagerInterceptorTests.cs]: What the point of this test? you dont check there was a reorg (using the interceptor)

Comment (fassadlr) [reply src/Stratis.Bitcoin.IntegrationTests/ConsensusManagerInterceptorTests.cs]: Have discussed on Slack 👍

File: src/Stratis.Bitcoin.IntegrationTests/ConsensusManagerInterceptorTests.cs (hunk 1 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.IntegrationTests/ConsensusManagerInterceptorTests.cs]: In this case I need to first create the node that I wish to disconnect from (minerA) so I first have to set up the nodes and then inject the interceptor function.

Also, I want to be specific on what the interceptor is intercepting, so thought it better to call `BlockDisconnectInterceptor`?

Thoughts?

Comment (dangershony) [reply src/Stratis.Bitcoin.IntegrationTests/ConsensusManagerInterceptorTests.cs]: ah yeah its going to be tricky so like this is good

File: src/Stratis.Bitcoin.IntegrationTests/ConsensusManagerInterceptorTests.cs (hunk 2 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.IntegrationTests/ConsensusManagerInterceptorTests.cs]: Good job very creative mate

**Pull request merged**